File: Distribute Coins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        function<pair<int, int>(TreeNode*)> explore;
        explore = [&explore] (TreeNode* node) -> pair<int, int> {
            int nett = node->val - 1, cost = 0;
            if (node->left != nullptr) {
                auto [left_nett, left_cost] = explore(node->left);
                cost +=left_cost + abs(left_nett);
                nett += left_nett;
            }

            if (node->right != nullptr) {
                auto [right_nett, right_cost] = explore(node->right);
                cost += right_cost + abs(right_nett);
                nett +=right_nett;
            }

            return make_pair(nett, cost);
        };
        return explore(root).second;
    }
};

File: Distribute Coins in Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func abs(n int) int {
    if n < 0 {
        return -n
    }
    return n
}
func distributeCoins(root *TreeNode) int {
    var explore func(*TreeNode)(int, int)
    explore = func(node *TreeNode)(int, int) {
        nett := node.Val - 1
        cost := 0
        if node.Left != nil {
            left_nett, left_cost := explore(node.Left)
            cost += left_cost + abs(left_nett)
            nett += left_nett
        }

        if node.Right != nil {
            right_nett, right_cost := explore(node.Right)
            cost += right_cost + abs(right_nett)
            nett += right_nett
        }

        return nett, cost
    }
    _, cost := explore(root)
    return cost
}
